go/chat/convsource.go

```diff
@@ -315,6 +315,68 @@ func (s *HybridConversationSource) identifyTLF(ctx context.Context, convID chat1
 	return nil
 }
 
+func (s *HybridConversationSource) resolveHoles(ctx context.Context, uid gregor1.UID,
+	thread *chat1.ThreadView, conv chat1.Conversation) (err error) {
+	defer s.Trace(ctx, func() error { return err }, "resolveHoles")()
+	var msgIDs []chat1.MessageID
+	// Gather all placeholder messages so we can go fetch them
+	for index, msg := range thread.Messages {
+		state, err := msg.State()
+		if err != nil {
+			continue
+		}
+		if state == chat1.MessageUnboxedState_PLACEHOLDER {
+			if index == len(thread.Messages)-1 {
+				// If the last message is a hole, we might not have fetched everything,
+				// so fail this case like a normal miss
+				return storage.MissError{}
+			}
+			msgIDs = append(msgIDs, msg.GetMessageID())
+		}
+	}
+	if len(msgIDs) == 0 {
+		// Nothing to do
+		return nil
+	}
+	if s.IsOffline() {
+		// Don't attempt if we are offline
+		return OfflineError{}
+	}
+
+	// Fetch all missing messages from server, and sub in the real ones into the placeholder slots
+	msgs, err := s.GetMessages(ctx, conv.GetConvID(), uid, msgIDs, conv.Metadata.FinalizeInfo)
+	if err != nil {
+		s.Debug(ctx, "resolveHoles: failed to get missing messages: %s", err.Error())
+		return err
+	}
+	msgLookup := make(map[chat1.MessageID]chat1.MessageUnboxed)
+	for _, msg := range msgs {
+		msgLookup[msg.GetMessageID()] = msg
+	}
+	for i, threadMsg := range thread.Messages {
+		state, err := threadMsg.State()
+		if err != nil {
+			continue
+		}
+		if state == chat1.MessageUnboxedState_PLACEHOLDER {
+			if msg, ok := msgLookup[threadMsg.GetMessageID()]; ok {
+				thread.Messages[i] = msg
+			} else {
+				s.Debug(ctx, "resolveHoles: did not fetch all placeholder messages, missing msgID: %d",
+					threadMsg.GetMessageID())
+				return fmt.Errorf("did not fetch all placeholder messages")
```

Comment:
I don't really know how this case could happen either, I just put it in there to be careful.